### DISCUSSION TITLE
Adding exiftool to a golem shiny package

### DISCUSSION THREAD

**Discussion by zcfogel-idfg** [ThinkR-open/golem#780] [Q&A]
I am in the process of converting an existing shiny app into a golem shiny package. The app is used for copying photos from SD cards to hard drives and renaming the photos once they have been copied to the hard drive. It renames them based on the photo UserLabel (the camera ID) and the date and time that the photo was taken. The app is run locally on the user’s computer rather than on a shiny server, and it has the user install Strawberry perl in order to utilize exiftool. It uses a combination of the exifr package and exiftool (run through a `system2()` call). The parts of the app that use exifr work fine when the app is in the form of a standard R project and after it was converted to a golem app. The part that runs exiftool through `system2()` work as an R project, but not in golem; it displays the error “`exiftool not found`.”

The segment using exiftool is currently written as such: 
`rename_cmnd <- paste("inst/app/www/exiftool -r -d %Y%m%d_%H%M%S \"-filename<",new_folder,"/${userlabel;s/ //g}/",vals$pcode,"_${userlabel;s/ //g}_${DateTimeOriginal}_${triggermode;s/(.)(.*\\s)(.)(.*)/$1$3/}${sequence;s/(\\d)(.*)/_$1/}.%e\" ",  input$selectedBatch, sep = "”)` where `input$selectedBatch`, `new_folder`, and `vals$pcode` are user-input values. Then, I actually call that command with `rcd<-system2(rename_cmnd)`. 

Vincent at ThinkR said "once your golem is installed, a file localised in  inst/app/www/exiftool  will be available with
`system.file(package="yourgolemname","app","www","exiftool")`." I am unsure where/when I call `system.file()`. Is it instead of `system2()`? Or before `system2()`? Or once I've loaded the app but before `run_app()`? 

Additionally, my package name (for now) is sdApp, and the R Project is called sd_package_v3.0; I assume that I do `package='sdApp'`?

Thank you!

**Comment by ColinFay** [comment]
Hey @zcfogel-idfg,

Thanks for your question. 

There are two sides to this question:

- `inst` is a folder that you can use while building a package, and the content of this folder will be copied __as is__ when the package is installed on a computer. 

- `system.file(package = "xyz")` is a base R function that can be called to refer to the  __folder where your {xyz}` package is installed on the machine__.

For example : 

```r
> system.file("img", package = "golem")
[1] "/Library/Frameworks/R.framework/Versions/4.1/Resources/library/golem/img"
```

Will refer to the `img` folder from `{golem}`: https://github.com/ThinkR-open/golem/tree/dev/inst/img
This folder is copied to your computer when you install the package. 

Also, if you do `browseURL(system.file(package = "golem"))`, you'll see that you get the R installation & the content of https://github.com/ThinkR-open/golem/tree/dev/inst, as is.

So you can __reliably__ refer to this folder on __every__ computer that has the package installed, hence you can read/write inside it. 

That being said, you should be aware that if you wish to send the package on the CRAN, that the CRAN doesn't allow packages that silently write to the user computer. 
Hence I'd suggest adding a way for the user to select the folder where they want to file to be written :) 
You can for example use https://github.com/thomasp85/shinyFiles

Best, 
C.
